File: ParkingLot/ParkingLot_Node/level.ts
import ParkingSpot from "./parkingSpot.ts";
import { Vehicle, VehicleType } from "./vehicles/vehicle.ts";

export default class Level {
  _floorNumber: number;
  _parkingSpots: ParkingSpot[];

  constructor(floorNumber: number, spotsCount: number) {
    this._floorNumber = floorNumber;
    this._parkingSpots = new Array(spotsCount);

    const bikeSpots = Math.floor(spotsCount * 0.6);
    for (let i = 0; i < bikeSpots; i++) {
      this._parkingSpots[i] = new ParkingSpot(i, VehicleType.bike);
    }

    const carSpots = Math.floor(spotsCount * 0.3);
    for (let i = bikeSpots; i < bikeSpots + carSpots; i++) {
      this._parkingSpots[i] = new ParkingSpot(i, VehicleType.car);
    }

    for (let i = bikeSpots + carSpots; i < spotsCount; i++) {
      this._parkingSpots[i] = new ParkingSpot(i, VehicleType.truck);
    }
  }

  get floorNumber() {
    return this._floorNumber;
  }

  get parkingSpots() {
    return this._parkingSpots;
  }

  parkVehicle(v: Vehicle) {
    const n = this._parkingSpots.length;

    for (let i = 0; i < n; i++) {
      const parkingSpot = this._parkingSpots[i];

      if (
        parkingSpot.isAvailable() &&
        parkingSpot.vehicleType === v.vehicleType
      ) {
        parkingSpot.parkVehicle(v);
        return true;
      }
    }
    return false;
  }

  unParkVehicle(v: Vehicle) {
    const n = this._parkingSpots.length;

    for (let i = 0; i < n; i++) {
      const parkingSpot = this._parkingSpots[i];

      if (parkingSpot.vehicle === v) {
        parkingSpot.unParkVehicle();
        return true;
      }
    }
    return false;
  }

  checkAvailability() {
    const n = this._parkingSpots.length;
    let count = 0;

    for (let i = 0; i < n; i++) {
      const parkingSpot = this._parkingSpots[i];

      if (parkingSpot.isAvailable()) {
        count++;
      }
    }

    return count;
  }
}

File: ParkingLot/ParkingLot_Go/level/level.go
package level

import (
	"parkingLot/parkingSpot"
	"parkingLot/vehicles"
	"sync"
)

func GetNewLevel(floor, spotsCount int) *Level {
	parkingSpots := make([]*parkingSpot.ParkingSpot, spotsCount)

	carSpots := int(float32(spotsCount) * 0.6)
	for i := 0; i < carSpots; i++ {
		parkingSpots[i] = parkingSpot.GetNewParkingSpot(i, vehicles.Car)
	}

	bikeSpots := int(float32(spotsCount) * 0.3)
	for i := carSpots; i < carSpots+bikeSpots; i++ {
		parkingSpots[i] = parkingSpot.GetNewParkingSpot(i, vehicles.Bike)
	}

	for i := carSpots + bikeSpots; i < spotsCount; i++ {
		parkingSpots[i] = parkingSpot.GetNewParkingSpot(i, vehicles.Truck)
	}

	return &Level{
		floor:        floor,
		parkingSpots: parkingSpots,
	}
}

type Level struct {
	floor        int
	parkingSpots []*parkingSpot.ParkingSpot
	mu           sync.Mutex
}

func (l *Level) GetId() int {
	return l.floor
}

func (l *Level) ParkVehicle(v vehicles.Vehicle) bool {
	for _, spot := range l.parkingSpots {
		if spot.GetType() == v.GetType() && spot.IsAvailable(&l.mu) {
			spot.ParkVehicle(v, &l.mu)
			return true
		}
	}
	return false
}

func (l *Level) UnParkVehicle(v vehicles.Vehicle) bool {
	for _, spot := range l.parkingSpots {
		if spot.GetVehicle() != nil && spot.GetVehicle().GetLicense() == v.GetLicense() {
			spot.UnParkVehicle(&l.mu)
			return true
		}
	}
	return false
}

func (l *Level) CheckAvailability() int {
	count := 0
	for _, spot := range l.parkingSpots {
		if spot.IsAvailable(&l.mu) {
			count++
		}
	}
	return count
}
